### PR TITLE
feat(merge): SeekingMerger — RocksDB-style dual loser trees (#222)

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Implements RocksDB-style merging iterator (**Option A** from #222 design exploration). Two independent `LoserTree` instances (forward min + backward max) over `MergeSource` — a new seekable-iter trait — for bidirectional iteration with init-time cross-direction migration. Refs #222.

This PR ships the **primitive layer**. Wiring `seek()` into the direction-switch path for sources with independent front/back cursors (LSM SST scanners, run readers) is the follow-up #280. Closing the small-fan-in perf regression vs `BinaryHeap` is the follow-up #283.

## Architecture

```
SeekingMerger<S: MergeSource>
  ├── forward_tree:  LoserTree<MergerEntry, EntryCmp>   (min, lazy-init)
  ├── backward_tree: LoserTree<MergerEntry, EntryCmp>   (max via reversed cmp, lazy-init)
  ├── sources: Vec<S>
  └── pending_forward_error / pending_backward_error: Option<Error>

MergeSource trait
  ├── fn next(&mut self) -> Option<IterItem>
  ├── fn next_back(&mut self) -> Option<IterItem>
  └── fn seek(&mut self, target: &InternalKey) -> Result<()>

CoherentMergeSource: MergeSource {}   // marker for shared-cursor sources
```

### Contracts that ship

| Surface | Semantic |
|---|---|
| `MergeSource::seek` | Independent-cursor sources MUST reposition both cursors per the per-method doc. Coherent-cursor sources (`CoherentMergeSource` marker) MAY no-op — their shared front/back cursor already enforces "no item yielded twice". |
| `DoubleEndedIterator for SeekingMerger<S>` | Type-bounded on `S: CoherentMergeSource`. Independent-cursor sources are type-system-rejected from `next_back()` calls at compile time until #280 wires seek into the switch path. |
| Init data preservation | If a source errors during init, sources that successfully prefetched earlier are NOT lost — their values stay in the tournament. The error is queued and surfaces on the next call. |
| Init cross-direction migration | If one direction's init pulls `None` from an already-exhausted source while the OPPOSITE tree still holds a buffered leaf for that source, the leaf is migrated (via `LoserTree::take_slot`) into the init-vec. Prevents silent data loss for cases like single source `[a, z]` (`next` yields `a` and prefetches `z`; `next_back` would otherwise return `None`). |
| Error-first contract | Queued errors (from init OR refill, either direction) surface at the top of the very next `next()` / `next_back()` call — they're not buried behind unrelated yields from other sources. |
| First-error-wins | `pending_*_error` is single-slot (`get_or_insert`). Multiple errors collapse to the first; the rest are discarded. |
| Cross-direction surfacing | A forward refill error surfaces on the next `next_back()` call too (and vice versa) so a caller can't sidestep it by switching direction. |

## Bench (criterion --quick, release)

```
N    MergeHeap   SeekingMerger    Δ
─────────────────────────────────────────
 2   17.7 µs     21.0 µs          +18% slower
 4   42.4 µs     49.5 µs          +17% slower
 8   99.2 µs     114.2 µs         +15% slower
16   247.4 µs    262.3 µs         +6%  slower
30   610.0 µs    569.0 µs         -7%  FASTER
```

Crossover at N≈20. SeekingMerger wins on large fan-in (compaction, level merges); BinaryHeap wins on small fan-in (typical L0 read). The remaining small-N gap is structural (LoserTree per-step constants > BinaryHeap on tight loops) and tracked as **#283** with concrete avenues (aggressive inlining, drop `Option` wrap via parallel `present` bitmap, SoA leaf layout, SIMD comparator).

Bench now reports P50/P99/P999 tail latency via `b.iter_custom` alongside criterion's mean+CI report.

## What ships

- `src/loser_tree.rs` — generic `LoserTree<E, F>` (Knuth tournament tree) with `build` / `peek_min` / `replace_min` / `pop_min` / `take_slot` / `winner_slot` / `slots` / `active_count` / `is_empty`. 11 unit tests. `take_slot(i)` is the migration primitive — O(log cap) removal of an arbitrary leaf with tournament replay.
- `src/merge_source.rs` — `MergeSource` trait + `CoherentMergeSource: MergeSource` marker + `Box<dyn>` blanket impl + `CoherentIterSource` adapter (no-op seek for `alloc::vec::IntoIter`, `VecDeque`, `BTreeMap::range`). 5 unit tests.
- `src/seeking_merger.rs` — `SeekingMerger<S: MergeSource>` with 16 unit tests covering forward / backward / mixed / empty / single-source / init-error / refill-error / cross-direction-error / init-migration / error-first-ordering scenarios.
- `benches/merge.rs` — side-by-side `Merge {N} (MergeHeap)` vs `Merge {N} (SeekingMerger)` for `N ∈ {2, 4, 8, 16, 30}` with P50/P99/P999 reporting and zero-based percentile math.

`src/merge.rs` (`Merger` / `MergeHeap`-based) ships unchanged. SeekingMerger is an opt-in alternative until production callers migrate.

## Test plan

- [x] 16/16 SeekingMerger unit tests pass
- [x] 11/11 LoserTree unit tests pass
- [x] 5/5 MergeSource unit tests pass
- [x] Full workspace `cargo nextest run` clean — no regression to existing `Merger`
- [x] `cargo clippy --lib --tests --benches -- -D warnings` clean
- [x] Bench: SeekingMerger faster at N=30 (compaction), regression on N≤16 tracked separately as #283

## Follow-up work for #222 closure

Migrating production callers (`range.rs`, compaction, mvcc_stream) needs `MergeSource` impls for:

- `table::iter::Iter` (SST scanner — already has `seek_lower_inclusive` / `seek_upper_inclusive`)
- `RunReader` (multi-table run iterator)
- Memtable `range_internal().filter().map(Ok)` chain (needs a seekable filter adapter or push the filter into the source)

Plus closing #283 perf regression so SeekingMerger can be the unconditional default. Each piece is bounded; the algorithmic primitives are all there. Next PR.

## Related

- #222 — design exploration that selected this approach
- #280 — wire `MergeSource::seek` into SeekingMerger's direction-switch path
- #283 — close the small-N perf regression vs `BinaryHeap`

## Summary by CodeRabbit

* **New Features**
  * Added bidirectional merge iteration and an alternative merge strategy variant for more flexible merging behavior.

* **Performance**
  * Benchmarks enhanced to report tail-latency percentiles (P50/P99/P999) for better visibility into merge performance.

* **Tests**
  * Comprehensive unit tests added to validate merge behavior, edge cases, and error handling.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/coordinode-lsm-tree/pull/284?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: benches/merge.rs [hunk 1]
```diff
@@ -1,43 +1,144 @@
...
+        // Criterion can legitimately call iter_custom with iters == 0
+        // during early-warmup probing. Without this guard the
+        // percentile indexing below would underflow on an empty
+        // samples vec (n.max(1) returns 1 but samples[0] would OOB).
+        return Duration::ZERO;
+    }
+    // Lossless clamp: bound `iters` by MAX_SAMPLES as u64 first,
+    // then convert via try_from — avoids `iters as usize`
+    // truncation on 32-bit targets, lint-clean.
+    let cap = usize::try_from(iters.min(MAX_SAMPLES as u64)).unwrap_or(MAX_SAMPLES);
```
[the remark this answers is not in the document]
Comment (polaz) [reply benches/merge.rs]: @coderabbitai Fixed all 3 sites — added `const MAX_SAMPLES_U64: u64` as a raw-literal twin so reservoir math stays in u64 (only narrows to usize at the cap site via try_from), and replaced the other two `as u64` casts with `u64::try_from(...).unwrap_or(u64::MAX)`.
